File: app/src/main/java/com/nptel/courses/online/utility/ColorUtility.java
package com.nptel.courses.online.utility;

import android.graphics.Color;

import java.util.Random;

public class ColorUtility {


    public static String CHOCOLATE = "#D2691E";
    public static String CORAL = "#FF7F50";
    public static String CRIMSON = "#DC143C";
    public static String FOREST_GREEN = "#228B22";
    public static String GOLD = "#FFD700";
    public static String HOT_PINK = "#FF69B4";
    public static String LIGHT_CORAL = "#F08080";
    public static String LIGHT_SEA_GREEN = "#20B2AA";
    public static String LIME_GREEN = "#32CD32";
    public static String MEDIUM_PURPLE = "#9370DB";
    public static String ORANGE_RED = "#FF4500";
    public static String SLATE_BLUE = "#6A5ACD";
    public static String STEEL_BLUE = "#4682B4";
    public static String TOMATO = "#FF6347";
    public static String VIOLATE = "#EE82EE";

    public static int getRandomColor(int redBound, int greenBound, int blueBound) {
        Random random = new Random();
        int r, g, b;
        r = random.nextInt(redBound);
        g = random.nextInt(greenBound);
        b = random.nextInt(blueBound);
        return Color.rgb(r, g, b);
    }

    public static int getPrimaryRandomColor() {
        Random random = new Random();
        int ran = random.nextInt(21);
        switch (ran) {
            case 0:
                return Color.BLUE;
            case 1:
                return Color.CYAN;
            case 2:
                return Color.GREEN;
            case 3:
                return Color.MAGENTA;
            case 4:
                return Color.RED;
            case 5:
                return Color.YELLOW;
            case 6:
                return Color.parseColor(CHOCOLATE);
            case 7:
                return Color.parseColor(CORAL);
            case 8:
                return Color.parseColor(CRIMSON);
            case 9:
                return Color.parseColor(FOREST_GREEN);
            case 10:
                return Color.parseColor(GOLD);
            case 11:
                return Color.parseColor(HOT_PINK);
            case 12:
                return Color.parseColor(LIGHT_CORAL);
            case 13:
                return Color.parseColor(LIGHT_SEA_GREEN);
            case 14:
                return Color.parseColor(LIME_GREEN);
            case 15:
                return Color.parseColor(MEDIUM_PURPLE);
            case 16:
                return Color.parseColor(ORANGE_RED);
            case 17:
                return Color.parseColor(SLATE_BLUE);
            case 18:
                return Color.parseColor(STEEL_BLUE);
            case 19:
                return Color.parseColor(TOMATO);
            case 20:
                return Color.parseColor(VIOLATE);
            default:
                return 0;
        }
    }

    public static int[] getRandomColors(int red, int green, int blue, int size) {
        return getRandomColors(red, green, blue, size, null);
    }

    public static int[] getRandomColors(int red, int green, int blue, int size, int[] previousColors) {
        int[] colors = new int[size];
        if (previousColors != null) {
            System.arraycopy(previousColors, 0, colors, 0, previousColors.length);
        }
        int i = 0;
        while (i < size) {
            int newColor = getRandomColor(red, green, blue);
            if (checkIfArrayContainsValue(colors, newColor)) continue;
            colors[i] = newColor;
            i++;
        }
        return colors;
    }

    private static boolean checkIfArrayContainsValue(int[] array, int value) {
        for (int element : array)
            if (element == value)
                return true;
        return false;
    }

    public static int ARGB_To_RGB(int argb) {
        int alpha = Color.alpha(argb);
        int red = Color.red(argb);
        int green = Color.green(argb);
        int blue = Color.blue(argb);
        float fraction = (256 - (float) alpha) / 256;
        red = red + (int) ((256 - red) * fraction);
        green = green + (int) ((256 - green) * fraction);
        blue = blue + (int) ((256 - blue) * fraction);
        return Color.rgb(red, green, blue);
    }

    public static int getLighterVersion(int color) {
        String hexColor = String.format("#%06X", (0xFFFFFF & color));
        return Color.parseColor("#36" + hexColor.substring(hexColor.length() - 6));
    }

    public static int getLighterVersion(int color, String alpha) {
        String hexColor = String.format("#%06X", (0xFFFFFF & color));
        return Color.parseColor(alpha + hexColor.substring(hexColor.length() - 6));
    }
}

File: app/src/main/java/com/nptel/courses/online/utility/ColorUtility.kt
package com.nptel.courses.online.utility

import android.graphics.Color
import java.util.*

object ColorUtility {

    var CHOCOLATE = "#D2691E"
    var CORAL = "#FF7F50"
    var CRIMSON = "#DC143C"
    var FOREST_GREEN = "#228B22"
    var GOLD = "#FFD700"
    var HOT_PINK = "#FF69B4"
    var LIGHT_CORAL = "#F08080"
    var LIGHT_SEA_GREEN = "#20B2AA"
    var LIME_GREEN = "#32CD32"
    var MEDIUM_PURPLE = "#9370DB"
    var ORANGE_RED = "#FF4500"
    var SLATE_BLUE = "#6A5ACD"
    var STEEL_BLUE = "#4682B4"
    var TOMATO = "#FF6347"
    var VIOLATE = "#EE82EE"

    fun getRandomColor(redBound: Int, greenBound: Int, blueBound: Int, dark: Boolean): Int {
        val random = Random()

        // light -> max = bound, min = 0 -> (bound - 0 + 1) + 0
        //                                  (bound)
        // dark -> max = 256, min = 256 - bound -> (256 - 256 + bound + 1) + 256 - bound
        //                                          (bound + 1) + 256 - bound

        val red: Int =
            random.nextInt(if (dark) redBound + 1 else redBound) + if (dark) 256 - redBound else 0
        val green: Int =
            random.nextInt(if (dark) greenBound + 1 else greenBound) + if (dark) 256 - greenBound else 0
        val blue: Int =
            random.nextInt(if (dark) blueBound + 1 else blueBound) + if (dark) 256 - blueBound else 0


        return Color.rgb(red, green, blue)
    }

    val primaryRandomColor: Int
        get() {
            val random = Random()
            val ran = random.nextInt(21)
            return when (ran) {
                0 -> Color.BLUE
                1 -> Color.CYAN
                2 -> Color.GREEN
                3 -> Color.MAGENTA
                4 -> Color.RED
                5 -> Color.YELLOW
                6 -> Color.parseColor(CHOCOLATE)
                7 -> Color.parseColor(CORAL)
                8 -> Color.parseColor(CRIMSON)
                9 -> Color.parseColor(FOREST_GREEN)
                10 -> Color.parseColor(GOLD)
                11 -> Color.parseColor(HOT_PINK)
                12 -> Color.parseColor(LIGHT_CORAL)
                13 -> Color.parseColor(LIGHT_SEA_GREEN)
                14 -> Color.parseColor(LIME_GREEN)
                15 -> Color.parseColor(MEDIUM_PURPLE)
                16 -> Color.parseColor(ORANGE_RED)
                17 -> Color.parseColor(SLATE_BLUE)
                18 -> Color.parseColor(STEEL_BLUE)
                19 -> Color.parseColor(TOMATO)
                20 -> Color.parseColor(VIOLATE)
                else -> 0
            }
        }

    fun getRandomColors(bound: Int, size: Int, dark: Boolean = false): List<Int> {
        return getRandomColors(bound, bound, bound, size, dark)
    }

    fun getRandomColors(
        red: Int,
        green: Int,
        blue: Int,
        size: Int,
        dark: Boolean,
        previousColors: List<Int> = emptyList()
    ): List<Int> {
        val colors = mutableListOf<Int>()
        colors.addAll(previousColors)
        var i = 0
        while (colors.size < size) {
            val newColor = getRandomColor(red, green, blue, dark)
            if (colors.contains(newColor)) continue
            colors.add(newColor)
        }
        return colors
    }

    private fun checkIfArrayContainsValue(array: IntArray, value: Int): Boolean {
        for (element in array) if (element == value) return true
        return false
    }

    fun ARGB_To_RGB(argb: Int): Int {
        val alpha = Color.alpha(argb)
        var red = Color.red(argb)
        var green = Color.green(argb)
        var blue = Color.blue(argb)
        val fraction = (256 - alpha.toFloat()) / 256
        red = red + ((256 - red) * fraction).toInt()
        green = green + ((256 - green) * fraction).toInt()
        blue = blue + ((256 - blue) * fraction).toInt()
        return Color.rgb(red, green, blue)
    }

    fun getLighterVersion(color: Int): Int {
        val hexColor = String.format("#%06X", 0xFFFFFF and color)
        return Color.parseColor("#36" + hexColor.substring(hexColor.length - 6))
    }

    fun getLighterVersion(color: Int, alpha: String): Int {
        val hexColor = String.format("#%06X", 0xFFFFFF and color)
        return Color.parseColor(alpha + hexColor.substring(hexColor.length - 6))
    }
}